frigate/object_detection.py

```diff
@@ -22,46 +29,150 @@ def ReadLabelFile(file_path):
     return ret
 
 class PreppedQueueProcessor(threading.Thread):
-    def __init__(self, cameras, prepped_frame_queue):
+    def __init__(self, cameras, prepped_frame_queue, tf_args):
 
         threading.Thread.__init__(self)
         self.cameras = cameras
         self.prepped_frame_queue = prepped_frame_queue
         
+        # print("tf_args: " + str(tf_args))
         # Load the edgetpu engine and labels
-        self.engine = DetectionEngine(PATH_TO_CKPT)
-        self.labels = ReadLabelFile(PATH_TO_LABELS)
+        if tf_args != None:
+            tf_model = tf_args.get('model', None)
+            tf_labels = tf_args.get('labels', None)
+            self.data_dir = tf_args.get('data_dir', DATA_DIR)
+        else:
+            tf_model = PATH_TO_CKPT
+            tf_labels = PATH_TO_LABELS
+            self.data_dir = DATA_DIR
+        print("Loading Tensorflow model: " + str(tf_model))
+        self.engine = DetectionEngine(tf_model)
+        if tf_labels:
+            self.labels = ReadLabelFile(tf_labels)
+        else:
+            self.labels = None
+        yesterday = datetime.datetime.now() - datetime.timedelta(days=1)
+        self.last_saved_det_frame_time = yesterday
+        self.last_saved_non_det_frame_time = yesterday
 
     def run(self):
         # process queue...
         while True:
             frame = self.prepped_frame_queue.get()
 
             # Actual detection.
-            objects = self.engine.DetectWithInputTensor(frame['frame'], threshold=frame['region_threshold'], top_k=3)
+#            objects = self.engine.DetectWithInputTensor(frame['frame'], threshold=frame['region_threshold'], top_k=3)
             # print(self.engine.get_inference_time())
 
-            # parse and pass detected objects back to the camera
-            parsed_objects = []
-            for obj in objects:
-                box = obj.bounding_box.flatten().tolist()
-                parsed_objects.append({
-                            'frame_time': frame['frame_time'],
-                            'name': str(self.labels[obj.label_id]),
-                            'score': float(obj.score),
-                            'xmin': int((box[0] * frame['region_size']) + frame['region_x_offset']),
-                            'ymin': int((box[1] * frame['region_size']) + frame['region_y_offset']),
-                            'xmax': int((box[2] * frame['region_size']) + frame['region_x_offset']),
-                            'ymax': int((box[3] * frame['region_size']) + frame['region_y_offset'])
-                        })
-            self.cameras[frame['camera_name']].add_objects(parsed_objects)
+            try:
+                # print("Calling TF Coral to detect objects")
+                detected_objects = self.engine.DetectWithImage(
```

Comment:
Using `DetectWithImage` means that this thread is spending time resizing, cropping, and then converting this to a numpy array prior to detecting objects. All the time spent doing that is time that the Coral is not detecting objects, and will reduce the total possible speed of the Coral.

Reply:
Yes, that is correct. I was weighing convenience and performance here. The face model has a different input tensor shape (320x320 vs 300x300 for coco) which requires a bit more code to detect and resize. I am ok with writing the extra code, although I'm not sure if it will add significant performance benefit.

While we are on the topic of resizing. What are your thoughts on relaxing the square requirement for regions? My tests show that passing a rectangular region with preserved aspect ratio works well.

Reply:
It is a significant performance benefit when you are trying to run detection on more than 5 cameras with lots of regions and anything above 1fps. I have tried it both ways. Because the model is trained on squares, the only way you can maintain the aspect ratio is to crop the image. We can relax the requirement, but I would want to crop in the separate thread where resizing happens so this thread doesn't have to calculate it.